batcher/batcher/cardano_chain_operations.go

```diff
@@ -601,3 +602,193 @@ func getOutputs(
 func getMaxUtxoCount(config *cardano.CardanoChainConfig, prevUtxosCnt int) int {
 	return max(int(config.MaxUtxoCount)-prevUtxosCnt, 0) //nolint:gosec
 }
+
+func (cco *CardanoChainOperations) generatePolicyAndMultisig(
+	validatorsData []eth.ValidatorChainData) (*cardano.ApexPolicyScripts, *cardano.ApexAddresses, error) {
+	keyHashes, err := cardano.NewApexKeyHashes(validatorsData)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	policyScripts := cardano.NewApexPolicyScripts(keyHashes)
+
+	addresses, err := cardano.NewApexAddresses(cco.cardanoCliBinary, uint(cco.config.NetworkMagic), policyScripts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &policyScripts, &addresses, nil
+}
+
+// CreateValidatorSetChangeTx implements core.ChainOperations.
+func (cco *CardanoChainOperations) CreateValidatorSetChangeTx(ctx context.Context, chainID string, nextBatchID uint64,
+	bridgeSmartContract eth.IBridgeSmartContract,
+	validatorsKeys *validatorobserver.Validators) (*core.GeneratedBatchTxData, error) {
+	// get validators data
+	validatorsData, ok := validatorsKeys.Data[chainID]
+	if !ok {
+		return nil, fmt.Errorf("couldn't find keys for chain:%s", chainID)
+	}
+
+	// new validator set policy, multisig & fee address
+	_, newAddresses, err :=
+		cco.generatePolicyAndMultisig(validatorsData.Keys)
+	if err != nil {
+		return nil, err
+	}
+
+	// get active validator set from bridge smart contract
+	activeValidatorsData, err := cco.getValidatorsChainData(ctx, bridgeSmartContract, chainID)
+	if err != nil {
+		return nil, err
+	}
+
+	// active validator set policy, multisig & fee address
+	activePolicy, activeAddresses, err :=
+		cco.generatePolicyAndMultisig(activeValidatorsData)
+	if err != nil {
+		return nil, err
+	}
+
+	protocolParams, err := cco.txProvider.GetProtocolParameters(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// get filtered & limited utxos
+	multisigUtxos, feeUtxos, isFeeOnly, err := cco.getUTXOsForValidatorChange(
+		activeAddresses.Multisig.Payment, activeAddresses.Fee.Payment)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(feeUtxos) == 0 {
+		return &core.GeneratedBatchTxData{
+			BatchType: uint8(ValidatorSetFinal),
+			TxRaw:     []byte{},
+			TxHash:    "",
+		}, nil
+	}
+
+	metadata, err := common.MarshalMetadata(common.MetadataEncodingTypeJSON, common.BatchExecutedMetadata{
+		BridgingTxType: common.BridgingTxTypeBatchExecution,
+		BatchNonceID:   nextBatchID,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// last slot number from indexer
+	slotNumber, err := cco.getSlotNumber()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		txRaw  []byte
+		txHash string
+	)
+
+	if isFeeOnly {
+		// last transaction sends funds from active fee key to new fee key
```

Comment:
if there is only one fee utxo and amount of that utxo is less than `MinimalUtxoAmount*2` we have already finished